### PR TITLE
PR for feature #58 (access private datasets and be able to upsert and full replace datasets)

### PR DESCRIPTION
**Pull request by marks**
Please see issue (feature request) #58; This is the re-do of PR #63 since it was not correctly based on the `dev` branch

### CODE REVIEW

**Comment by marks** [comment]
I am seeing some tests fail locally but I think it is code not related to my changes in the `dev` branch. Please do let me know if this is not the case!

**Comment by dmpe** [comment]
https://travis-ci.org/Chicago/RSocrata/builds/88642764#L1424

File: R/writeData.R [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  }
+  
+  # TODO: error handling
+  # errorHandling(response)
+  
+  return(response)
+}
+
+
+#' @description Method for updating Socrata datasets
```
Comment (dmpe) [inline R/writeData.R]: title is missing

**Comment by marks** [comment]
How may I transmit the username/password used for the tests that require a Socrata login? I have been able to successfully test on my own travis-ci.org repo using environment variables.

Thanks!

**Comment by dmpe** [comment]
http://kieranhealy.org/blog/archives/2015/10/16/using-containerized-travis-ci-to-check-r-in-rmarkdown-files/ may help or again what i said in #58

**Comment by marks** [comment]
You can't use containerized Travis unless sudo isn't required.. I also think it seems an admin of the repo is best off just adding the environment variables from the Travis interface


--
Mark Silverberg
(m) 512 826 7004
http://twitter.com/skram

**Comment by dmpe** [comment]
oh, i forgot, sorry.


---------- Původní zpráva ----------

Od: Mark Silverberg <notifications@github.com>

Komu: Chicago/RSocrata <RSocrata@noreply.github.com>

Datum: 2. 11. 2015 16:21:35

Předmět: Re: [RSocrata] PR for feature #58 (access private datasets and be 
able to upsert and full replace datasets) (#64)


"
You can't use containerized Travis unless sudo isn't required.. I also think
it seems an admin of the repo is best off just adding the environment 
variables from the Travis interface





--

Mark Silverberg

(m) 512 826 7004

http://twitter.com/skram



> On Nov 2, 2015, at 10:19 AM, John Malc <notifications@github.com> wrote:

> 

> http://kieranhealy.org/blog/archives/2015/10/16/using-containerized-travis
-ci-to-check-r-in-rmarkdown-files/ may help or again what i said in #58

> 

> —

> Reply to this email directly or view it on GitHub.

> 


—

Reply to this email directly or view it on GitHub
(https://github.com/Chicago/RSocrata/pull/64#issuecomment-153050972). 





"=

**Comment by dmpe** [comment]
cc @tomschenkjr

**Comment by tomschenkjr** [comment]
Still reviewing this pull request. Some issues are showing-up in the process. Here are some quick notes so far:

* `read.socrata(url = "https://data.cityofchicago.org/Health-Human-Services/Food-Inspections/4ijn-s7e5")` generates an error because of the [`rowCount`](https://github.com/marks/RSocrata/blob/e4ee27081088664fc25e4a2d9bcd4d20d477f4ef/R/returnData.R#L116) line. Specifically, I'm trying to understand the grander concept around `getMetadata()`, which is trying to fetch a lot of information, but is using little of it. Does it make sense to just keep it simple to help troubleshooting?
* `read.socrata(url = "https://data.cityofchicago.org/Buildings/Building-Permits/ydr8-5enu")` is rather FUBAR, but this is an existing problem. For a reason I cannot determine yet, the list is not being properly transformed into a data frame. Instead, it's being transformed into an ugly list.

There are some other issues, but am looking at these two first.

**Comment by marks** [comment]
@tomschenkjr: I am seeing those errors as well but I am not confident they are part of this PR. I am getting those errors on the `dev` branch (which doesn't have this PR merged in)

Thanks,

Mark

**Comment by tomschenkjr** [comment]
Yeah, it's unrelated to the write functions you added so as an FYI at this point. Once I clear up those other issues I'll be looking at these new functions.

Sent by Outlook<http://taps.io/outlookmobile> for Android



On Mon, Nov 16, 2015 at 4:54 AM -0800, "Mark Silverberg" <notifications@github.com<mailto:notifications@github.com>> wrote:


@tomschenkjr<https://github.com/tomschenkjr>: I am seeing those errors as well but I am not confident they are part of this PR. I am getting those errors on the dev branch (which doesn't have this PR merged in)

Thanks,

Mark

-
Reply to this email directly or view it on GitHub<https://github.com/Chicago/RSocrata/pull/64#issuecomment-157020153>.


________________________________
This e-mail, and any attachments thereto, is intended only for use by the addressee(s) named herein and may contain legally privileged and/or confidential information. If you are not the intended recipient of this e-mail (or the person responsible for delivering this document to the intended recipient), you are hereby notified that any dissemination, distribution, printing or copying of this e-mail, and any attachment thereto, is strictly prohibited. If you have received this e-mail in error, please respond to the individual sending the message, and permanently delete the original and any copy of any e-mail and printout thereof.

**Comment by dmpe** [comment]
Any issue now can be profiled using https://github.com/rstudio/profvis. 
```
devtools::install_github("rstudio/profvis")
library(profvis)
library(RSocrata)

profvis({
  g <- read.socrata(url = "https://data.cityofchicago.org/Health-Human-Services/Food-Inspections/4ijn-s7e5")
  print(g)
})

profvis({
  b <- read.socrata(url = "https://data.cityofchicago.org/Buildings/Building-Permits/ydr8-5enu")
  print(b)
})
```
For the second issue e.g. like this : socratalist


> Specifically, I'm trying to understand the grander concept around getMetadata(), which is trying to fetch a lot of information, but is using little of it. 

Yes, there are 2 ways: Either you use execute another query on the desired dataset with =Count(*)  (getQueryRowCount does that) or you fetch that information from the socrata's undocumented ` columns.json view`. Here, @marks can provide us some help and tell what is better and more realible (I guess the first one, otherwise there would be no such method). In ither case, we can use the first option only and provide the second (from me) as just another function for end-users.

**Comment by marks** [comment]
@dmpe - i would suggest using the `COUNT(*)` as opposed to using columns.json. Documented metadata APIs coming soon!

**Comment by dmpe** [comment]
Thanks, one issue solved. Some code changes will be needed. Assumption must be made that if from me then at earliest in Christmas break. 

With the second issue, @marks  can also try help us. (Not so helpful and not so nicely summarized and I am sorry for it.) Do you know why do some datasets (and btw. some not; see above) have an "invalid" json file, which may cause our issues. (well, I guess these were attempts to fix them: #19 #3 ) 

```
library(jsonlite)
df<-jsonlite::fromJSON("https://data.cityofchicago.org/resource/ydr8-5enu.json")
```

Cannot test it now, but as i remember it was failing at least some time ago.

**Comment by marks** [comment]
@dmpe - could you help me understand this issue by sharing a list of datasets that are giving invalid JSON? I am very confident all JSON that we produce and put out fits the standard but it might not be exactly what `jsonlite`/`R` is expecting. I just need a little more info to be able to look into this more with the team.

**Comment by dmpe** [comment]
Yeeh, code from #3 is blocked so i cannot test it anymore. 
At least these are working, now. Most probably an error in our code. 
```
df<-jsonlite::fromJSON("https://data.cityofchicago.org/resource/ydr8-5enu.json")
df1<-jsonlite::fromJSON("https://data.cityofchicago.org/resource/kn9c-c2s2.json")
df1<-jsonlite::fromJSON("https://data.cityofchicago.org/resource/4ijn-s7e5.json")
dfs1<-jsonlite::fromJSON("http://data.undp.org/resource/wxub-qc5k.json")
```

**Comment by dmpe** [comment]
After spending huge amount of time on it, I start to think the error is somewhere from httr or curl package. 
Infact, when I run this:
```
gM3 <- getMetadata(url = "https://data.cityofchicago.org/resource/6zsd-86xi.json")
 Show Traceback
 
 Rerun with Debug
 Error in curl::curl_fetch_memory(url, handle = handle) : 
  Couldn't resolve host name 

```
But the same code for 
```
gM2 <- getMetadata(url = "https://data.cityofboston.gov/resource/awu8-dc52")
```
is not an issue. WTF is it ?

**Comment by marks** [comment]
@dmpe @tomschenkjr is there anything I can do to help keep this moving forward?

Appreciate all your hard work!

**Comment by tomschenkjr** [comment]
I think we have a bit of scope creep in this PR, but I'm going to continue to evaluate the parts of it to clean-up the errors. Hope to do it this week.


Tom Schenk Jr.

Chief Data Officer

Department of Innovation and Technology

City of Chicago

(312) 744-2770

tom.schenk@cityofchicago.org

data.cityofchicago.org


________________________________

**Pull request closed without merging**